### DISCUSSION TITLE
Seeking advice for handling `java.lang.VerifyError` with java.util.concurrent (a.k.a executor) instrumentation

### DISCUSSION THREAD

**Discussion by chris-pinola-rf** [open-telemetry/opentelemetry-java-instrumentation#4496] [Q&A] [answered]
Hi folks,

I'm fairly new to the JVM ecosystem, but I'm attempting to instrument a variety of Kotlin and Java applications with the OpenTelemetry agent.

The applications that I'm instrumenting also use the Rollout.io/CloudBees Feature Management ([docs](https://docs.cloudbees.com/docs/cloudbees-feature-management/latest/getting-started/java-sdk), [Maven package](https://mvnrepository.com/artifact/io.rollout.rox/rox-java-server/5.0.5)) SDK. The Rollout SDK uses `java.util.concurrent` to perform asynchronous network I/O. Therefore the OpenTelemetry agent's `executor` module is injecting some bytecode into those library functions to facilitate context propagation.

The problem is that when I enable the OpenTelemetry agent on these applications, the Rollout SDK fails to initialize and throws a [`java.lang.VerifyError`](https://docs.oracle.com/javase/7/docs/api/java/lang/VerifyError.html). From what I can tell, for security purposes, the JVM attempts to verify library bytecode before executing it. Since the agent has injected bytecode into some of the `java.util.concurrent` classes, that library bytecode no longer matches the JVM's "known good" version for said classes and throws this error at runtime.

I've identified two possible workarounds:

1. Disable the `java.util.concurrent` instrumentation by setting the `OTEL_INSTRUMENTATION_EXECUTOR_ENABLED` environment variable to `false`.
2. Disable the JVM's "verifier" by setting the `JDK_JAVA_OPTIONS` environment variable to `-noverify`.

Both workarounds, however, have significant downsides. We have some code that also uses `java.util.concurrent` and would benefit from the `executor` instrumentation module. Also, my gut tells me that disabling verification is going to open the door to some potentially dangerous/difficult to debug issues.

So, I'm looking for advice on what to do. Are there other alternatives that I haven't discovered? Have I completely misdiagnosed the situation? I would greatly appreciate any suggestions. Thanks!

**Comment by trask** [comment, marked as answer]
hey @chris-pinola-rf! it seems like a bug in the javaagent (the goal is definitely for this kind of thing not to happen 😅), is there any more detail that is printed out with the `VerifyError`?

**Reply by chris-pinola-rf** [reply]
Yes, here are some more details from the exception:
```
java.util.concurrent.ExecutionException: java.lang.VerifyError: Inconsistent stackmap frames at branch target 38
Exception Details:
  Location:
    io/rollout/okio/AsyncTimeout$a.run()V @38: nop
  Reason:
    Type top (current frame, locals[2]) is not assignable to 'io/rollout/okio/AsyncTimeout$a' (stack map, locals[2])
  Current Frame:
    bci: @56
    flags: { }
    locals: { 'io/rollout/okio/AsyncTimeout$a', 'io/opentelemetry/javaagent/shaded/io/opentelemetry/context/Scope', top, 'io/rollout/okio/AsyncTimeout', 'java/lang/Class' }
    stack: { }
  Stackmap Frame:
    bci: @38
    flags: { }
    locals: { 'io/rollout/okio/AsyncTimeout$a', 'io/opentelemetry/javaagent/shaded/io/opentelemetry/context/Scope', 'io/rollout/okio/AsyncTimeout$a' }
    stack: { }
  Bytecode:
    0000000: 122d 122f b800 354c 2b2a b800 3ba7 0016
    0000010: 123d b800 435f 1245 5fb9 004b 0300 a700
    0000020: 0457 014c 2a4d 0012 0259 3a04 c2b8 0009
    0000030: 594e c700 0919 04c3 a7ff ee2d b200 08a6
    0000040: 000d 01b3 0008 1904 c3a7 001c 1904 c3a7
    0000050: 000b 3a05 1904 c319 05bf 2db6 000a a7ff
    0000060: c857 a7ff c401 4da7 0004 4d2b c600 092b
    0000070: b900 5001 00a7 0015 123d b800 435f 1245
    0000080: 5fb9 004b 0300 a700 0457 002c c600 052c
    0000090: bfb1                                   
  Exception Handler Table:
    bci [45, 56] => handler: 82
    bci [59, 73] => handler: 82
    bci [76, 79] => handler: 82
    bci [82, 87] => handler: 82
    bci [38, 56] => handler: 97
    bci [59, 73] => handler: 97
    bci [76, 94] => handler: 97
    bci [0, 16] => handler: 16
    bci [38, 101] => handler: 106
    bci [107, 120] => handler: 120
    bci [16, 30] => handler: 33
    bci [120, 134] => handler: 137
  Stackmap Table:
    same_locals_1_stack_item_frame(@16,Object[#7])
    same_locals_1_stack_item_frame(@33,Object[#7])
    same_frame(@34)
    same_locals_1_stack_item_frame(@35,Object[#77])
    append_frame(@36,Object[#77])
    append_frame(@38,Object[#3])
    chop_frame(@39,1)
    append_frame(@59,Top,Object[#2],Object[#4])
    same_frame(@76)
    full_frame(@82,{Object[#3],Object[#77],Top,Top,Object[#4]},{Object[#7]})
    full_frame(@90,{Object[#3],Object[#77],Top,Object[#2]},{})
    full_frame(@97,{Object[#3],Object[#77]},{Object[#5]})
    full_frame(@101,{Object[#3],Object[#77]},{})
    same_locals_1_stack_item_frame(@106,Object[#7])
    append_frame(@107,Object[#7])
    same_frame(@117)
    same_locals_1_stack_item_frame(@120,Object[#7])
    same_locals_1_stack_item_frame(@137,Object[#7])
    same_frame(@138)
    same_frame(@139)
    same_frame(@145)

    at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
    at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
    at io.rollout.client.Core$1.a(SourceFile:94)
    at io.rollout.client.Core$1.call(SourceFile:90)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
    at java.base/java.lang.Thread.run(Thread.java:832)
```

**Reply by mateuszrzeszutek** [reply]
Hey @chris-pinola-rf 
Asking just to make sure: which version of the javaagent are you using?

**Reply by chris-pinola-rf** [reply]
This exception was observed from an application that was running version 1.6.2 of the javaagent, but I can reproduce this exact error on 1.7.0 too.

**Reply by mateuszrzeszutek** [reply]
Looks like a fresh bug then. Can you [create a bug issue](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/new/choose) for this? Please include the JVM version and vendor, javaagent version, rollout library version, etc.
It'd be very helpful if you could provide a link to a repro scenario; a simple app that causes this issue.
Thanks!

**Reply by chris-pinola-rf** [reply]
@mateuszrzeszutek Okay, I'll try to have that to you later today. 👍

**Reply by chris-pinola-rf** [reply]
I've filed a bug issue https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/4511 which also links to the [repro repo](https://github.com/chris-pinola-rf/otel_cloudbees_repro). Please let me know if I can provide any more info.
